server/api/migrations/versions/67b0169ba6b4_add_alert_history.py

```diff
@@ -0,0 +1,107 @@
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+"""add alert history
+
+Revision ID: 67b0169ba6b4
+Revises: fcf2ea01f99a
+Create Date: 2024-10-30 12:01:35.944309
+
+"""
+
+import os
+
+import sqlalchemy as sa
+from alembic import op
+from sqlalchemy.dialects import mysql
+
+# revision identifiers, used by Alembic.
+revision = "67b0169ba6b4"
+down_revision = "fcf2ea01f99a"
+branch_labels = None
+depends_on = None
+
+
+def upgrade():
+    # ### commands auto generated by Alembic - please adjust! ###
+    op.create_table(
+        "alert_history",
+        sa.Column("id", sa.Integer(), nullable=False),
+        sa.Column(
+            "activation_time", mysql.DATETIME(timezone=True, fsp=3), nullable=False
+        ),
+        sa.Column(
+            "name", sa.String(length=255, collation="utf8mb3_bin"), nullable=False
+        ),
+        sa.Column(
+            "project", sa.String(length=255, collation="utf8mb3_bin"), nullable=False
+        ),
+        sa.Column("data", sa.JSON(), nullable=True),
+        sa.Column(
+            "entity_id", sa.String(length=255, collation="utf8mb3_bin"), nullable=False
+        ),
+        sa.Column(
+            "entity_kind",
+            sa.String(length=255, collation="utf8mb3_bin"),
+            nullable=False,
+        ),
+        sa.Column(
+            "event_kind", sa.String(length=255, collation="utf8mb3_bin"), nullable=False
+        ),
+        sa.Column(
+            "severity", sa.String(length=255, collation="utf8mb3_bin"), nullable=False
+        ),
+        sa.Column("number_of_events", sa.Integer(), nullable=False),
+        sa.PrimaryKeyConstraint("activation_time", "id", name="_alert_history_uc"),
+    )
+    op.create_index(
+        "ix_alert_history_entity_id", "alert_history", ["entity_id"], unique=False
+    )
+    op.create_index(
+        "ix_alert_history_entity_kind", "alert_history", ["entity_kind"], unique=False
+    )
+    op.create_index(
+        "ix_alert_history_event_kind", "alert_history", ["event_kind"], unique=False
+    )
+    op.create_index("ix_alert_history_name", "alert_history", ["name"], unique=False)
+    op.create_index(
+        "ix_alert_history_project", "alert_history", ["project"], unique=False
+    )
+    op.create_index(
+        "ix_alert_history_severity", "alert_history", ["severity"], unique=False
+    )
+    # ### end Alembic commands ###
+
+    # Add initial partitioning
+    partition_interval = os.getenv("PARTITION_INTERVAL", "YEARWEEK")
+    default_partition_value = os.getenv("DEFAULT_PARTITION_VALUE", "2023-01-01")
+    partition_sql = f"""
+         ALTER TABLE alert_history
+         PARTITION BY RANGE ({partition_interval}(activation_time)) (
+             PARTITION p0 VALUES LESS THAN ({partition_interval}('{default_partition_value}'))
+         );
```

Comment:
This depends on values retrieved from outside, so at a minimum you should verify that both `partition_interval` and `default_partition_value` don't contain special chars and only contain valid chars for SQL.
The other reason (even though less of a concern here due to being a specific installation flow) is that this is subject to SQL-injection exploits. Many static-analysis tools will catch this code and warn against it. Ideally other ways should be used to inject parameters into SQL statements executed.

Reply:
@theSaarco agreed 💯 , I've rewritten the logic.